spec/client/Rollup-Integration-guide.md

```diff
@@ -0,0 +1,201 @@
+    consensusState = ConsensusState(header.appHash, header.time)
+
+    setConsensusState(clientMessage.clientId, clientMessage.height, consensusState)
+
+    // create mapping between consensus state and the current time for fraud proof waiting period
+    setSubmitTime(consensusState, currentTime())
+}
+```
+
+### `checkForMisbehaviour`
+
+Misbehaviour verification has a different purpose for rollup architectures than it does in traditional consensus mechanisms.
+
+Typical consensus mechanisms, like proof-of-stake, are self-reliant on ordering. Thus, we must have mechanisms to detect when the consensus set is violating the ordering rules. For example, in cometBFT, the misbehaviour verification checks that header times are monotonically increasing and that there exists only one valid header for each height.
+
+However, with rollups the ordering is derived from the data availability layer. Thus, even if there is a consensus violation in the rollup consensus, it can be resolved by the DA layer and the consensus rules of the rollup. E.g. even if the sequencer signs multiple blocks at the same height, the canonical block is the first block submitted to the DA layer.
+
+Thus, so long as the verification method encodes the consensus rules of the rollup architecture correctly (for instance, ensuring the header submitted is the earliest one for the given height), then there is no need to verify misbehaviour of the rollup consensus. The consensus is derived from the DA layer, and so if the DA client is frozen due to misbehaviour, this should halt proof verification in the rollup client as well.
+
+Instead, the misbehaviour most relevant for rollups is in the application layer, as the transactions are executed by the sequencer but not by the underlying data availability layer. For ZK rollups, the application is already proven correct so there is no need for application misbehaviour verification. However, optimistic rollups must provide the ability for off-chain processes to submit a proof that the application hash submitted in the header was the result of an incorrect computation of transaction(s) in the block i.e. a fraud proof.
+
+The optimistic fraud proof verifier should be implemented as a smart contract. Since the fraud prover is dependent not on the consensus but on the application state machine itself. Thus each rollup instance needs its own fraud prover. Having each fraud prover encoded directly in the client requires a different implementation for each rollup instance. Instead, calling out to a separate smart contract allows the client to be reused for all instances, and for new fraud provers to be uploaded for a new rollup application.
+
+```typescript
+// optimistic rollup fraud proof
+// the misbehaviour must be associated with a height on the rollup
+function checkForMisbehaviour(clientMessage: ClientMessage) {
+    // unmarshalling logic ommitted
+    clientId = clientMessage.clientId
+    misbehaviour = clientMessage.misbehaviour
```

Comment:
Just to confirm: there are no checks we need to perform on `clientMessage.header` for misbehaviour detection then?